include/beast/static_string.hpp

```diff
@@ -0,0 +1,696 @@
+    }
+
+    /// Returns the number of characters that can be held in currently allocated storage.
+    size_type
+    capacity() const
+    {
+        return N;
+    }
+    
+    /// Reduces memory usage by freeing unused memory.
+    void
+    shrink_to_fit()
+    {
+        // no-op
+    }
+
+    /// Clears the contents.
+    void
+    clear()
+    {
+        resize(0);
+    }
+
+    /** Changes the number of characters stored.
+
+        @note No value-initialization is performed.
+    */
+    void
+    resize(std::size_t n);
+
+    /** Changes the number of characters stored.
+
+        If the resulting string is larger, the new
+        characters are initialized to the value of `c`.
+    */
+    void
+    resize(std::size_t n, CharT c);
+
+    /// Compare two character sequences.
+    template<std::size_t M>
+    int compare(static_string<M, CharT, Traits> const& rhs) const;
+
+    /// Return the characters as a `basic_string`.
+    std::basic_string<CharT, Traits>
+    to_string() const
+    {
+        return std::basic_string<
+            CharT, Traits>{&s_[0], n_};
+    }
+
+private:
+    void
+    assign(CharT const* s);
+};
+
+template<std::size_t N, class CharT, class Traits>
+static_string<N, CharT, Traits>::
+static_string()
+    : n_(0)
+{
+    s_[0] = 0;
+}
+
+template<std::size_t N, class CharT, class Traits>
+static_string<N, CharT, Traits>::
+static_string(static_string const& s)
+    : n_(s.n_)
+{
+    Traits::copy(&s_[0], &s.s_[0], n_ + 1);
+}
+
+template<std::size_t N, class CharT, class Traits>
+template<std::size_t M>
+static_string<N, CharT, Traits>::
+static_string(static_string<M, CharT, Traits> const& s)
+{
+    if(s.size() > N)
+        throw std::length_error("static_string overflow");
+    n_ = s.size();
+    Traits::copy(&s_[0], &s.s_[0], n_ + 1);
+}
+
+template<std::size_t N, class CharT, class Traits>
+auto
+static_string<N, CharT, Traits>::
+operator=(static_string const& s) ->
+    static_string&
+{
+    n_ = s.n_;
+    Traits::copy(&s_[0], &s.s_[0], n_ + 1);
+    return *this;
+}
+
+template<std::size_t N, class CharT, class Traits>
+template<std::size_t M>
+auto
+static_string<N, CharT, Traits>::
+operator=(static_string<M, CharT, Traits> const& s) ->
+    static_string&
+{
+    if(s.size() > N)
+        throw std::length_error("static_string overflow");
+    n_ = s.size();
+    Traits::copy(&s_[0], &s.s_[0], n_ + 1);
+    return *this;
+}
+
+template<std::size_t N, class CharT, class Traits>
+template<std::size_t M>
+static_string<N, CharT, Traits>::
+static_string(const CharT (&s)[M])
+    : n_(M-1)
+{
+    static_assert(M-1 <= N,
+        "static_string overflow");
+    Traits::copy(&s_[0], &s[0], M);
+}
+
+template<std::size_t N, class CharT, class Traits>
+template<std::size_t M>
+auto
+static_string<N, CharT, Traits>::
+operator=(const CharT (&s)[M]) ->
+    static_string&
+{
+    static_assert(M-1 <= N,
+        "static_string overflow");
+    n_ = M-1;
+    std::copy(&s[0], &s[M], &s_[0]);
```

Comment:
You use `Traits::copy(&s_[0], &s[0], M);` everywhere in the file except here. Is there a reason, or should this one be changed for consistency.